### ISSUE TITLE
Regex group support

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#2093]
For some regex use cases, a user may need to go beyond just knowing whether there was a match and actually extract the strings that matched. An example implementation of this is [Python regex grouping](https://docs.python.org/3/howto/regex.html#grouping).

As a specific example use case, this could be a possible alternative to the implementation envisioned for #2092. Instead of using functions to split/slice desired fields, and re-join back into a string, a regex approach in Python could look like:

```
$ python
>>> import re
>>> query='www.brimsecurity.com'
>>> m = re.search(r'(\w+\.\w+)$', query)
>>> print(m.group(1))
brimsecurity.com
```

**Comment by philrz** [comment]
A community zync user inquired about this functionality. In their own words:

> Do you have any idea how we can use Zed to extract a substring based on a regular expression ? Suppose I need to extract the street number out of a postal address, or maybe  extract a particular pieces of a URL, or parse a special text field in a CSV file.
> Would it be possible to enhance the `split` function using a regular expression pattern `split(s: string, <pattern>) -> [string]` and returning an array of matching substrings? Perhaps something similar to TCL `regexp ?switches? exp string ?matchVar? ?subMatch1 ... subMatchN?`?

**Comment by philrz** [comment]
Verified in Zed commit e9624c3.

Performing the Zed equivalent of the Python example shown in the opening text of this issue:

```
$ zq -version
Version: v1.2.0-85-ge9624c3b

$ echo '{"query": "www.brimdata.io"}' | zq -z 'yield regexp("(\\w+\\.\\w+)$", query)' -
["brimdata.io","brimdata.io"]
```

A more sophisticated example is shown in the [`regexp()` docs](https://zed.brimdata.io/docs/next/language/functions/regexp).

Notice how we had to escape the backslashes, which is because the regexp is passed to the function as a string. #4148 has been opened to track a possible improvement on this.

Thanks @mattnibs!
